double_linked_list/double_linked_list.c

```diff
@@ -3,22 +3,34 @@
 #include <stdbool.h>
 #include "double_linked_list.h"
 
-list* list_create(){ //how to tell if malloc worked?
+list* list_create(){
     list* l = (list*)malloc(sizeof(list));
+    if (l == NULL)
+        return NULL;
     l->quantity = 0;
     l->head = NULL;
     l->tail = NULL;
     return l;
 }
 
-void list_free(list* l){
-    node* first = l->head;
-    while (first->next) {
-            node *second = first;
-            first = first->next;
-            free(second);
+void list_free(list* l) {
+    if (l->head == NULL)
+        free(l);
+    else {
+        node *first = l->head;
+        if (l->quantity == 1)
+            free(first);
+        else if (l->quantity == 0)
+            return;
```

Comment:
Imho this is too complex. Also both `if (l->head == NULL)` and `else` paths does the same thing - `free(l)`. Sometimes such redundancy is okay, but here it is totally redundant. I wonder if a smart compiler would change this to:
```C
if (l->head != NULL) { ... }
free(l);
```
...but probably not.

In the end I would rewrite the `list_free` function to:
```C
void list_free(list* l) {
    if (l->quantity > 0) {
        node* current = l->head;
        node* next = l->head->next;
        
        while(next) {
            free(current);
            current = next;
            next = next->next;
        }
        
        // deallocating last element
        free(current);
    }
    
    free(l);
}
```